protocol/blockrelay/blockrelay.go

```diff
@@ -0,0 +1,274 @@
+	}
+
+	requestQueue := []*daghash.Hash{inv.Hash}
+	requestQueueSet := map[daghash.Hash]struct{}{
+		*inv.Hash: {},
+	}
+	pendingBlocks := map[daghash.Hash]struct{}{}
+
+	// In case the function closes earlier than expected, we wanna make sure requestedBlocks is
+	// clean from any pending blocks.
+	defer deleteFromRequestedBlocks(pendingBlocks)
+	for len(requestQueue) > 0 {
+		shouldStop, err := requestBlocks(netAdapter, router, peer, msgChan, dag, &invsQueue,
+			&requestQueue, requestQueueSet)
+		if err != nil {
+			return false, err
+		}
+		if shouldStop {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func readInv(msgChan <-chan wire.Message,
+	invsQueue *[]*wire.MsgInvRelayBlock) (inv *wire.MsgInvRelayBlock, shouldStop bool, err error) {
+
+	if len(*invsQueue) > 0 {
+		inv, *invsQueue = (*invsQueue)[0], (*invsQueue)[1:]
+		return inv, false, nil
+	}
+
+	for {
+		msg, isOpen := <-msgChan
+		if !isOpen {
+			return nil, true, nil
+		}
+
+		inv, ok := msg.(*wire.MsgInvRelayBlock)
+		if ok {
+			return inv, false, nil
+		}
+
+		return nil, false, errors.Errorf("unrequested %s message in the block relay flow", msg.Command())
+	}
+}
+
+// readMsgBlock returns the next msgBlock in msgChan, and populates invsQueue in any inv messages that arrives between.
+//
+// Note: this function assumes msgChan can contain only wire.MsgInvRelayBlock and wire.MsgBlock messages.
+func readMsgBlock(msgChan <-chan wire.Message,
+	invsQueue *[]*wire.MsgInvRelayBlock) (msgBlock *wire.MsgBlock, shouldStop bool, err error) {
+
+	for {
+		const stallResponseTimeout = 30 * time.Second
+		select {
+		case <-time.After(stallResponseTimeout):
+			return nil, false, errors.Errorf("stalled for %s", stallResponseTimeout)
+		case msg, isOpen := <-msgChan:
+			if !isOpen {
+				return nil, true, nil
+			}
+
+			inv, ok := msg.(*wire.MsgInvRelayBlock)
+			if !ok {
+				return msg.(*wire.MsgBlock), false, nil
+			}
+
+			*invsQueue = append(*invsQueue, inv)
+		}
+	}
+}
+
+func deleteFromRequestedBlocks(blockHashes map[daghash.Hash]struct{}) {
+	for hash := range blockHashes {
+		hash := hash
+		requestedBlocks.delete(&hash)
+	}
+}
+
+func requestBlocks(netAdapater *netadapter.NetAdapter, router *netadapter.Router, peer *peerpkg.Peer, msgChan <-chan wire.Message,
+	dag *blockdag.BlockDAG, invsQueue *[]*wire.MsgInvRelayBlock, requestQueue *[]*daghash.Hash,
+	requestQueueSet map[daghash.Hash]struct{}) (shouldStop bool, err error) {
+
+	var hashesToRequest []*daghash.Hash
+	if len(*requestQueue) > wire.MsgGetRelayBlocksHashes {
+		hashesToRequest, *requestQueue = (*requestQueue)[:wire.MsgGetRelayBlocksHashes],
+			(*requestQueue)[wire.MsgGetRelayBlocksHashes:]
+	} else {
+		hashesToRequest, *requestQueue = *requestQueue, nil
+	}
+
+	pendingBlocks := map[daghash.Hash]struct{}{}
+	for _, hash := range hashesToRequest {
+		delete(requestQueueSet, *hash)
+		pendingBlocks[*hash] = struct{}{}
+		exists := requestedBlocks.addIfExists(hash)
+		if exists {
+			return false, nil
```

Comment:
Do we really want to exit the function here? What about the rest of the hashesToRequest?